src/items/appItemHooks.ts

```diff
@@ -0,0 +1,143 @@
+  iFrameRef: React.RefObject<HTMLIFrameElement>;
+  requestApiAccessToken: (payload: {
+    id: UUID;
+    app: string;
+    origin: string;
+  }) => Promise<{
+    token: Token;
+  }>;
+}): void => {
+  const [channel, setChannel] = useState<MessageChannel>();
+
+  useEffect(() => {
+    // receive message from app through MessageChannel
+    const onMessage = async (e: MessageEvent): Promise<void> => {
+      const { data, origin: requestOrigin } = e;
+
+      const POST_MESSAGE_KEYS = buildPostMessageKeys(item.id);
+
+      // responds only to corresponding app
+      if (!appUrl?.includes(requestOrigin)) {
+        return;
+      }
+
+      const { type, payload } = JSON.parse(data);
+
+      switch (type) {
+        case POST_MESSAGE_KEYS.GET_AUTH_TOKEN: {
+          // eslint-disable-next-line no-unused-expressions
+          channel?.port1.postMessage(
+            JSON.stringify({
+              type: POST_MESSAGE_KEYS.GET_AUTH_TOKEN_SUCCESS,
+              payload: {
+                token: await requestApiAccessToken({ id: item.id, ...payload }),
+              },
+            }),
+          );
+          break;
+        }
+
+        case POST_MESSAGE_KEYS.POST_AUTO_RESIZE: {
+          // item should not be manually resizable
+          if (item.settings.isResizable) {
+            return;
+          }
+          // iframe must be mounted
+          if (iFrameRef.current === null) {
+            return;
+          }
+          // payload should be number
+          if (typeof payload !== 'number') {
+            return;
+          }
+          iFrameRef.current.height = payload.toString();
+          break;
+        }
+      }
+    };
+
+    const windowOnMessage = (e: MessageEvent): void => {
+      const { data, origin: requestOrigin } = e;
+
+      const POST_MESSAGE_KEYS = buildPostMessageKeys(item.id);
+
+      // responds only to corresponding app
+      if (!appUrl?.includes(requestOrigin)) {
+        return;
+      }
+
+      // return context data and message channel port to app
+      const { type } = JSON.parse(data);
+      if (type === POST_MESSAGE_KEYS.GET_CONTEXT) {
+        // create/reset channel and
+        // Listen for messages on port1
+        const channel = new MessageChannel();
+        const { port1 } = channel;
+        setChannel(channel);
+        port1.onmessage = onMessage;
+
+        // Transfer port2 to the iframe
+        // provide port2 to app and item's data
+        // eslint-disable-next-line no-unused-expressions
+        iFrameRef?.current?.contentWindow?.postMessage(
+          JSON.stringify({
+            type: POST_MESSAGE_KEYS.GET_CONTEXT_SUCCESS,
+            payload: {
+              apiHost,
+              itemId: item.id,
+              settings: item.settings,
+              memberId: member?.id,
+              permission,
+              lang: item.settings?.lang || member?.extra?.lang || DEFAULT_LANG,
+              context,
+            },
+          }),
+          '*',
```

Comment:
This should definitely be set to the target app origin. See https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#syntax at `targetOrigin`. Can this be set to `new URL(appUrl).origin`?

Reply:
I will try.

Reply:
It works, the messages target our domain now, i.e. `apps.dev.graasp.org`